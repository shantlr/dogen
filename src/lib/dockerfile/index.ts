import { DockerfileOp, DockerfileTarget } from './types';

const formatOp = (op: DockerfileOp): string => {
  switch (op.type) {
    case 'WORKDIR': {
      return `WORKDIR ${op.value}`;
    }
    case 'LABEL': {
      return `LABEL ${op.name}=${op.value}`;
    }
    case 'EXPOSE': {
      return `EXPOSE ${op.port}${op.protocol ? `/${op.protocol}` : ''}`;
    }
    case 'ENV': {
      return `ENV ${op.name}=${op.value}`;
    }
    case 'COPY': {
      const params = [];
      if (typeof op.from === 'string') {
        params.push(`--from=${op.from}`);
      }

      params.push(op.src);
      if (op.dst) {
        params.push(op.dst);
      }
      return `COPY ${params.join(' ')}`;
    }
    case 'CMD': {
      if (typeof op.cmd === 'string') {
        return `CMD ${op.cmd}`;
      }
      return `CMD [${op.cmd.map((c) => `"${c}"`).join(',')}]`;
    }
    case 'RUN': {
      if (typeof op.cmd === 'string') {
        return `RUN ${op.cmd}`;
      }
      return `RUN [${op.cmd.map((c) => `"${c}"`).join(',')}]`;
    }
    default:
  }
  throw new Error(`Unknown dockerfile op: ${JSON.stringify(op)}`);
};

const formatDockerfileTarget = (target: DockerfileTarget) => {
  const res: string[] = [
    `FROM ${
      typeof target.from === 'string' ? target.from : target.from.as
    } AS ${target.as}`,
    ...target.ops.map((o) => formatOp(o)),
  ];

  return res.join('\n');
};

export const formatDockerfile = (targets: DockerfileTarget[]): string => {
  return targets.map((t) => formatDockerfileTarget(t)).join('\n\n');
};
