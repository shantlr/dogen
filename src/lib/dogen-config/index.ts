import { readFile } from 'fs/promises';
import path from 'path';

import { cloneDeep, mergeWith } from 'lodash';
import * as v from 'valibot';

import { addDefaultInPlace } from '../core/presets/utils/add-defaults';
import { isFileExists } from '../utils';

import { Appendable, DeepOptional } from './types';

export const CONFIG_FILE_NAMES = [
  { name: '.dogenrc' },
  { name: '.dogenrc.json' },
];

const copyFromSchema = v.object({
  from: v.string(),
  src: v.string(),
  dst: v.string(),
});

const baseSchema = v.object({
  node: v.optional(
    v.object({
      from: v.optional(v.string()),
      setupPackageManagerVersion: v.optional(v.boolean()),
    }),
  ),
  target_prefix: v.optional(v.string()),

  package_manager: v.optional(v.picklist(['npm', 'yarn@1', 'yarn@4'])),

  container: v.optional(
    v.object({
      workdir: v.optional(v.string()),
    }),
  ),

  extract_deps: v.optional(
    v.object({
      jq_image_name: v.optional(v.string()),
      target_name: v.optional(v.string()),
    }),
  ),

  install: v.optional(
    v.object({
      from: v.optional(v.string()),
      target_name: v.optional(v.string()),
      // keep_cache: v.optional(v.boolean()),
      // npmrc: v.optional(v.union(v.string(), v.boolean())),
      cmd: v.optional(v.string()),
    }),
  ),

  build: v.optional(
    v.object({
      target_name: v.optional(v.string()),
      cmd: v.nullish(v.string()),
      script: v.nullish(v.string()),

      output_dir: v.nullish(v.string()),

      src_files: v.nullish(v.array(v.string())),
      src_copy_from: v.nullish(v.array(copyFromSchema)),
      src_additional_files: v.nullish(v.array(v.string())),
      src_detect_additional_files: v.nullish(
        v.array(
          v.union([
            v.string(),
            v.object({
              oneOf: v.array(v.string()),
            }),
          ]),
        ),
      ),

      post_files: v.nullish(v.array(v.string())),
      post_additional_files: v.nullish(v.array(v.string())),
      post_detect_additional_files: v.nullish(v.array(v.string())),
      post_copy_from: v.nullish(v.array(copyFromSchema)),
    }),
  ),

  run: v.optional(
    v.object({
      target_name: v.optional(v.string()),
      cmd: v.optional(v.string()),
      script: v.optional(v.string()),
      expose: v.optional(v.number()),
    }),
  ),

  serve: v.optional(
    v.object({
      from_image: v.optional(v.string()),
      config: v.optional(v.string()),
      config_dst: v.optional(v.string()),

      target_name: v.optional(v.string()),
      cmd: v.optional(v.string()),
      script: v.optional(v.string()),
      expose: v.optional(v.number()),
    }),
  ),
});

export const DOGEN_DEFAULT_CONFIG = {
  node_image: 'node:22-alpine',
  node: {
    from: 'node:22-alpine',
    setupPackageManagerVersion: true,
  },
  container: {
    workdir: '/app',
  },
  extract_deps: {
    target_name: 'extract_deps',
  },
  install: {
    target_name: 'install_modules',
  },
  build: {
    target_name: 'build',
  },
  run: {
    target_name: 'service',
  },
  serve: {
    target_name: 'serve',
  },
};

export type DogenConfig = v.InferOutput<typeof baseSchema>;
export type DogenConfigDeepOptional = DeepOptional<DogenConfig>;
export type DogenConfigAppend = Appendable<DogenConfig>;

export type PackageManagerName = NonNullable<DogenConfig['package_manager']>;

export type ExtendMutateConfig = {
  default?: DogenConfigDeepOptional;
  append?: DogenConfigAppend;
};

const assertDogenConfig = async (content: unknown) => {
  const warnings: string[] = [];

  const parsed = v.parse(baseSchema, content);
  // if (parsed.build?.cmd && parsed.build?.script) {
  //   warnings.push(
  //     `config 'build.script' will be ignored: both 'build.cmd' and 'build.script' has been provided`,
  //   );
  // }
  // if (parsed.run?.cmd && parsed.run?.script) {
  //   warnings.push(
  //     `config 'run.script' will be ignored: both 'run.cmd' and 'run.script' has been provided`,
  //   );
  // }

  return {
    warnings,
    parsed,
  };
};

const parseDogenConfig = async (str: string) => {
  try {
    const content = JSON.parse(str);

    const { parsed, warnings } = await assertDogenConfig(content);
    return {
      config: parsed,
      warnings,
    };
  } catch (err) {
    if (err instanceof SyntaxError) {
      throw new Error(`Failed to parse JSON: ${err.message}`);
    }

    throw err;
  }
};

export const getDogenConfig = async (dir: string) => {
  for (const c of CONFIG_FILE_NAMES) {
    const p = path.resolve(dir, c.name);
    if (await isFileExists(p)) {
      const res = await parseDogenConfig((await readFile(p)).toString());
      return {
        warnings: res.warnings,
        config: res.config,
        isDefault: false as const,
        configPath: p,
      };
    }
  }

  return {
    warnings: [] as string[],
    isDefault: true as const,
    configPath: null,
  };
};

export const applyConfigExtensions = (
  config: DogenConfig | undefined,
  extensions: (ExtendMutateConfig | undefined | null)[],
): DogenConfig => {
  if (!extensions.length) {
    return config ?? {};
  }

  const value = cloneDeep(config ?? {});
  for (const extension of extensions) {
    if (extension?.default) {
      mergeWith(value, extension.default, (objValue, srcValue) => {
        if (
          (typeof objValue !== 'object' && objValue != null) ||
          Array.isArray(objValue) ||
          objValue instanceof Date
        ) {
          return objValue;
        }
      });
    }
    if (extension?.append) {
      appendValuesInPlace(value, extension.append);
    }
  }

  return value;
};

function appendValuesInPlace<T>(
  val: T,
  append: Appendable<T> | undefined | null,
) {
  if (!append) {
    return val;
  }

  if (typeof append === 'object' && append) {
    for (const [key, value] of Object.entries(append)) {
      if (Array.isArray(value)) {
        if (Array.isArray(val[key as keyof T])) {
          (val[key as keyof T] as unknown[]).push(...value);
        } else if (val[key as keyof T] === undefined) {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          val[key as keyof T] = [...value];
        }
      } else if (value && typeof value === 'object') {
        if (val[key as keyof T] === undefined) {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          val[key as keyof T] = {};
        }
        appendValuesInPlace(val[key as keyof T], value);
      }
    }
  }
}

export const mergeConfigExtensions = (
  extensions: (ExtendMutateConfig | undefined)[],
) => {
  const res: ExtendMutateConfig = {
    default: {},
    append: {},
  };

  for (const extension of extensions) {
    if (extension?.default) {
      addDefaultInPlace(res.default!, extension.default);
    }

    if (extension?.append) {
      appendValuesInPlace(res.append, extension.append);
    }
  }

  return res;
};
