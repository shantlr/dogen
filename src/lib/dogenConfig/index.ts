import path from 'path';
import { isFileExists } from '../utils';
import { readFile } from 'fs/promises';
import { DogenInputConfig } from '../types';

const CONFIG_FILE_NAMES = [{ name: '.dogenrc' }];

const assertDogenConfig = (content: DogenInputConfig) => {
  const warnings: string[] = [];
  if (content.build?.cmd && content.build?.script) {
    warnings.push(
      `config 'build.script' will be ignored: both 'build.cmd' and 'build.script' has been provided`
    );
  }
  if (content.run?.cmd && content.run?.script) {
    warnings.push(
      `config 'run.script' will be ignored: both 'run.cmd' and 'run.script' has been provided`
    );
  }

  return {
    warnings,
  };
};

export const parseDogenConfig = (str: string) => {
  const content = JSON.parse(str);
  const { warnings } = assertDogenConfig(content);
  return {
    config: content as DogenInputConfig,
    warnings,
  };
};

export const detectDogenConfig = async (dir: string) => {
  for (const c of CONFIG_FILE_NAMES) {
    const p = path.resolve(dir, c.name);
    if (await isFileExists(p)) {
      const str = (await readFile(p)).toString();
      return parseDogenConfig(str);
    }
  }

  return {
    warnings: [],
    config: null,
  };
};
