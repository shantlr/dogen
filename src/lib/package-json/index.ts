import { readFile } from 'fs/promises';
import path from 'path';
import { isFileExists } from '../utils';

export interface PackageJson {
  name: string;
  version: string;
  main?: string;

  scripts?: Record<string, string>;

  dependencies?: Record<string, string>;
  devDependencies?: Record<string, string>;
}

export const findPackageJson = async (
  dir: string,
  root: string
): Promise<{
  dir: string;
  packageJson: PackageJson;
}> => {
  const p = path.resolve(dir, 'package.json');
  if (await isFileExists(p)) {
    return {
      dir,
      packageJson: await readFile(p)
        .then((r) => r.toString())
        .then((r) => JSON.parse(r) as PackageJson),
    };
  }

  const parent = path.resolve(dir, '..');
  if (parent !== dir && parent.startsWith(root)) {
    return findPackageJson(parent, root);
  }

  throw new Error('package.json not found');
};

export const packageHasDependency = (
  pkg: Pick<PackageJson, 'dependencies' | 'devDependencies'>,
  dep: string | string[]
): boolean => {
  if (typeof dep === 'string') {
    return Boolean(pkg.dependencies?.[dep] || pkg.devDependencies?.[dep]);
  }
  return dep.some((d) =>
    Boolean(pkg.dependencies?.[d] || pkg.devDependencies?.[d])
  );
};
