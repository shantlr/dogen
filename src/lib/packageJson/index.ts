import { readFile, stat } from 'fs/promises';
import path from 'path';

export interface PackageJson {
  name: string;
  version: string;

  scripts?: Record<string, string>;

  dependencies?: Record<string, string>;
  devDependencies?: Record<string, string>;
}

export const findPackageJson = async (
  dir: string,
  root: string
): Promise<{
  dir: string;
  packageJson: PackageJson;
}> => {
  try {
    const p = path.resolve(dir, 'package.json');
    await stat(p);
    return {
      dir,
      packageJson: await readFile(p)
        .then((r) => r.toString())
        .then((r) => JSON.parse(r) as PackageJson),
    };
  } catch (err) {
    if (err?.code === 'ENOENT') {
      const parent = path.resolve(dir, '..');
      if (parent !== dir && parent.startsWith(root)) {
        return findPackageJson(parent, root);
      }

      throw new Error('package.json not found');
    }
    throw err;
  }
};
