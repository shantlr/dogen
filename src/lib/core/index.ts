import { DepGraph } from 'dependency-graph';
import { DockerfileTarget } from '../dockerfile/types';
import { PackageJson } from '../package-json';
import { AnyPreset } from './presets/create-preset';
import { keyBy } from 'lodash';

/**
 * fill graph with presets dependencies
 * used to determine targets definition order and presets that are expected to be included in final dockerfile
 */
const fillPresetDepsGraph = ({
  graph,
  node,
  allPresets,
}: {
  graph: DepGraph<AnyPreset>;
  node: AnyPreset;
  allPresets: Record<string, AnyPreset>;
}) => {
  if (graph.hasNode(node.name)) {
    return graph;
  }

  graph.addNode(node.name, node);
  node.includes.forEach((include) => {
    if (!(include in allPresets)) {
      throw new Error(
        `preset '${node.name}' require '${include}' but not found`
      );
    }
    fillPresetDepsGraph({
      graph,
      node: allPresets[include],
      allPresets,
    });
    graph.addDependency(node.name, include);
  });
};
const resolvePresetsToInclude = ({
  root,
  allPresets,
}: {
  root: AnyPreset;
  allPresets: Record<string, AnyPreset>;
}) => {
  const graph = new DepGraph<AnyPreset>();
  fillPresetDepsGraph({ graph, node: root, allPresets });

  const order = graph.overallOrder();
  const presets = order.map((c) => graph.getNodeData(c));

  const targetAliases: Record<string, string> = {};
  presets.forEach((preset) => {
    for (const [target] of Object.entries(preset.targets)) {
      targetAliases[`@${preset.name}/${target}`] = '';
    }
  });

  return {
    presets,
    targetAliases,
  };
};

/**
 * Replace target ref (starting with '@') with resolved name
 */
const replaceTargetsRef = (
  target: DockerfileTarget,
  aliases: Record<string, string>,
  opt?: { assertInvalidRef?: boolean }
) => {
  if (typeof target.from === 'string') {
    if (target.from in aliases) {
      target.from = aliases[target.from];
    } else if (opt?.assertInvalidRef && target.from.startsWith('@')) {
      throw new Error(
        `Could not find target ${target.from} used in '${target.as}'`
      );
    }
  }

  target.ops.forEach((op) => {
    if (op.type === 'COPY' && typeof op.from === 'string') {
      if (op.from in aliases) {
        op.from = aliases[op.from];
      } else if (opt?.assertInvalidRef && op.from.startsWith('@')) {
        throw new Error(
          `Could not find target '${op.from}' used in '${target.as}'`
        );
      }
    }
  });
};

export const presetToDockerfileTargets = async (
  usedPreset: AnyPreset,
  opt: {
    config: Record<string, any>;
    projectDir: string;
    packageJson: PackageJson;
    includable: AnyPreset[];
  }
) => {
  const { presets, targetAliases } = resolvePresetsToInclude({
    root: usedPreset,
    allPresets: keyBy([usedPreset, ...opt.includable], 'name') as Record<
      string,
      AnyPreset
    >,
  });

  const res: DockerfileTarget[] = [];
  for (const preset of presets) {
    const config = preset.mapConfig ? await preset.mapConfig(opt) : opt.config;
    const input = {
      config,
      inputConfig: opt.config,
      projectDir: opt.projectDir,
      packageJson: opt.packageJson,
    };

    const localTargetAliases = {};
    const presetTargets: DockerfileTarget[] = [];

    for (const [key, target] of Object.entries(preset.targets)) {
      const t: DockerfileTarget = {
        comment: target.comment,
        as: typeof target.as === 'function' ? target.as(input) : target.as,
        from:
          typeof target.from === 'function' ? target.from(input) : target.from,
        ops: (
          (typeof target.ops === 'function'
            ? await target.ops(input)
            : target.ops) ?? []
        ).filter((o) => o),
      };
      presetTargets.push(t);
      localTargetAliases[`@/${key}`] = t.as;
      targetAliases[`@${preset.name}/${key}`] = t.as;
    }
    presetTargets.forEach((target) => {
      replaceTargetsRef(target, localTargetAliases);
    });

    res.push(...presetTargets);
  }

  res.forEach((target) => {
    replaceTargetsRef(target, targetAliases, {
      assertInvalidRef: true,
    });
  });

  return res;
};
