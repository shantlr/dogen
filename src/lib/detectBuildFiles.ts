import { PackageJson } from './packageJson';
import { DogenResolvedConfig } from './types';
import { filter, flatMap, map } from 'lodash';
import path from 'path';
import { isPathExists } from './utils';

export const depsFiles: {
  [depName: string]: string[];
} = {
  typescript: ['tsconfig.json'],
  vite: ['public', 'vite.config.ts', 'vite.config.js', 'tsconfig.node.json'],
  'react-scripts': ['public'],
  prettier: ['.prettierrc'],
  eslint: [
    '.eslintrc',
    '.eslintrc.js',
    '.eslintrc.cjs',
    '.eslintrc.yaml',
    '.eslintrc.yml',
    '.eslintrc.json',
  ],
};

const defaultIncludes = ['src'];

const resolveFiles = async (
  relativePaths: string[],
  dir: string,
  excludes: string[]
) => {
  const res = await Promise.all(
    relativePaths.map(async (rel) => {
      const p = path.resolve(dir, rel);
      if (!excludes.some((e) => e.startsWith(p)) && (await isPathExists(p))) {
        return p;
      }
      return null;
    })
  );
  return filter(res);
};

export const detectBuildFiles = async ({
  config,
  dir,
  packageJson,
}: {
  dir: string;
  packageJson: PackageJson;
  config: Pick<DogenResolvedConfig, 'build'>;
}) => {
  // TODO: refacto
  // const files = await readdir(dir, { withFileTypes: true });
  const res: string[] = [];

  const excludes = (
    (typeof config.build?.excludes === 'string'
      ? [config.build.excludes]
      : config.build?.excludes) || []
  ).map((e) => path.resolve(dir, e));

  if (config.build?.includes) {
    const includes =
      typeof config.build.includes === 'string'
        ? [config.build.includes]
        : config.build.includes;

    res.push(
      ...includes
        .map((i) => path.resolve(dir, i))
        .filter((i) => !excludes.some((e) => !e.startsWith(i)))
    );
  } else {
    const includes = await Promise.all([
      resolveFiles(defaultIncludes, dir, excludes),
      ...map(depsFiles, async (depFiles, depName) => {
        if (
          packageJson.dependencies?.[depName] ||
          packageJson.devDependencies?.[depName]
        ) {
          return resolveFiles(depFiles, dir, excludes);
        }
        return [];
      }),
    ]);
    res.push(...flatMap(includes));
  }

  // const excludes = Array.isArray(config.build?.excludes)
  //   ? config.build?.excludes
  //   : config.build?.excludes
  //   ? [config.build?.excludes]
  //   : [];

  // if (config.build?.includes) {
  //   // provided build files
  //   const includes =
  //     typeof config.build.includes === 'string'
  //       ? [config.build.includes]
  //       : config.build.includes;
  //   res.push(...includes.filter((i) => !excludes.includes(i)));
  // } else {
  //   // auto detect build files
  //   let includes = ['src'];

  //   forEach(depsFiles, (files, depName) => {
  //     if (
  //       packageJson.devDependencies?.[depName] ||
  //       packageJson?.dependencies?.[depName]
  //     ) {
  //       includes.push(...files);
  //     }
  //   });

  //   includes = uniq(includes);

  //   includes.forEach((i) => {
  //     if (!excludes.includes(i) && files.find((f) => f.name === i)) {
  //       res.push(i);
  //     }
  //   });
  // }

  //#region Extra files
  if (Array.isArray(config.build?.extraIncludes)) {
    config.build.extraIncludes.forEach((f) => {
      if (!res.includes(f)) {
        res.push(f);
      }
    });
  } else if (config.build?.extraIncludes) {
    if (!res.includes(config.build?.extraIncludes)) {
      res.push(config.build?.extraIncludes);
    }
  }
  //#endregion

  return res;
};
