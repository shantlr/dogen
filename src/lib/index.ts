import path from 'path';
import { findPackageJson } from './packageJson';
import { appendFile, writeFile } from 'fs/promises';
import { formatDockerfile } from './dockerfile';
import { isFileExists } from './utils';
import { detectDogenConfig } from './dogenConfig';
import { presetToDockerfileTargets } from './core';
import { Preset } from './core/presets/createPreset';
import { allPresets, defaulPresets } from './core/presets';

export const generateDockerfile = async ({
  dir = process.cwd(),
  rootDir = process.env.HOME ?? '/',
  dockerfileConflict,
  config: customConfig,
  presets = defaulPresets,
  mapConfig,
  includablePresets = allPresets,
}: {
  dir?: string;
  rootDir?: string;
  dockerfileConflict?: 'overwrite' | 'append';

  presets?: Preset<any, any, any>[];
  includablePresets?: Preset<any, any, any>[];
  /**
   * If not provided, will try to autodetect it
   */
  config?: Record<string, any>;
  /**
   * map provided config or autodetected config
   */
  mapConfig?: (
    detectedConfig: Record<string, any>
  ) => Record<string, any> | Promise<Record<string, any>>;
}) => {
  const { dir: projectDir, packageJson } = await findPackageJson(dir, rootDir);

  let inputConfig = customConfig;
  //#region Autodetect config if not provided
  if (!inputConfig) {
    const { config, warnings } = await detectDogenConfig(projectDir);
    inputConfig = config;
    if (warnings.length) {
      console.log(warnings.map((w) => `WARN: ${w}`).join('\n'));
    }
  }
  //#endregion

  const selectedPreset = presets.find((p) =>
    p.shouldUsePreset({ projectDir, packageJson })
  );
  if (!selectedPreset) {
    throw new Error(`NO_COMPATIBLE_PRESET`);
  }

  const config =
    typeof mapConfig === 'function'
      ? await mapConfig(inputConfig)
      : inputConfig;

  const targets = await presetToDockerfileTargets(selectedPreset, {
    config: config ?? {},
    projectDir,
    packageJson,
    includable: includablePresets,
  });
  const dockerfile = formatDockerfile(targets);

  let operation: 'created' | 'updated';

  //#region Write/Append Dockerfile
  const dockerfilePath = path.resolve(projectDir, 'Dockerfile');
  const alreadyExists = await isFileExists(dockerfilePath);

  const content = `#dogen\n${dockerfile}\n#enddogen\n`;

  if (!alreadyExists || dockerfileConflict === 'overwrite') {
    await writeFile(dockerfilePath, content);
    if (alreadyExists) {
      operation = 'updated';
    } else {
      operation = 'created';
    }
  } else if (dockerfileConflict === 'append') {
    operation = 'updated';
    await appendFile(dockerfilePath, `\n${content}`);
  } else {
    throw new Error(`DOCKERFILE_ALREADY_EXISTS`);
  }
  //#endregion

  return {
    operation,
    dockerfilePath,
  };
};
