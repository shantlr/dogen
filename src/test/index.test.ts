import { vol } from 'memfs';
import { generateDockerfile } from '../lib';
import { readFile } from 'fs/promises';

jest.mock('fs/promises');

describe('generateDockerfile', () => {
  beforeEach(() => {
    vol.reset();
    jest.restoreAllMocks();
  });

  it('should generate Dockerfile for service using yarn', async () => {
    const packageJson = {
      name: 'express-server',
      version: '1.0.0',
      main: 'build/index.js',
    };
    vol.fromJSON(
      {
        '/service/package.json': JSON.stringify(packageJson),
        '/service/yarn.lock': '',
      },
      '/tmp'
    );
    const spy = jest.spyOn(process, 'cwd');
    spy.mockReturnValue('/tmp');

    await generateDockerfile({
      dir: '/service',
    });
    const dockerfile = (await readFile('/service/Dockerfile')).toString();

    expect(dockerfile).toMatchSnapshot();
  });

  it('should follow config install keep cache', async () => {
    const packageJson = {
      name: 'express-server',
      version: '1.0.0',
      main: 'build/index.js',
    };
    const dogenrc = {
      install: {
        keepCache: true,
      },
    };
    vol.fromJSON(
      {
        '/service/package.json': JSON.stringify(packageJson),
        '/service/yarn.lock': '',
        '/service/.dogenrc': JSON.stringify(dogenrc),
      },
      '/tmp'
    );

    await generateDockerfile({
      dir: '/service',
    });
    const dockerfile = (await readFile('/service/Dockerfile')).toString();

    expect(dockerfile).toMatchSnapshot();
  });

  it('should follow config mount npmrc', async () => {
    const packageJson = {
      name: 'express-server',
      version: '1.0.0',
      main: 'build/index.js',
    };
    const dogenrc = {
      install: {
        npmrc: true,
      },
    };
    vol.fromJSON(
      {
        '/service/package.json': JSON.stringify(packageJson),
        '/service/yarn.lock': '',
        '/service/.dogenrc': JSON.stringify(dogenrc),
      },
      '/tmp'
    );

    await generateDockerfile({
      dir: '/service',
    });
    const dockerfile = (await readFile('/service/Dockerfile')).toString();

    expect(dockerfile).toMatchSnapshot();
  });
});
