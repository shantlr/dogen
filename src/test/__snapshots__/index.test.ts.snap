// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDockerfile should follow config install keep cache 1`] = `
"#dogen
FROM alpine:3.12 AS jq
RUN apk add --update --no-cache jq

FROM node:latest AS node-base

# Extract minimal fields for dependencies installation
# This step avoid reinstalling node_modules due to field that is unrelated
FROM jq AS extract-pkg-deps
COPY package.json /tmp/package.json
RUN jq '{name, dependencies, devDependencies}' < /tmp/package.json > /tmp/deps.json

# Install node_modules
FROM node-base AS install-modules
WORKDIR /service
COPY --from=extract-pkg-deps /tmp/deps.json package.json
COPY yarn.lock yarn.lock
RUN yarn install --pure-lockfile --non-interactive

# Build
FROM node-base AS build
WORKDIR /service
COPY --from=install-modules /service/node_modules node_modules
COPY package.json package.json
CMD yarn build
#enddogen
"
`;

exports[`generateDockerfile should follow config mount npmrc 1`] = `
"#dogen
FROM alpine:3.12 AS jq
RUN apk add --update --no-cache jq

FROM node:latest AS node-base

# Extract minimal fields for dependencies installation
# This step avoid reinstalling node_modules due to field that is unrelated
FROM jq AS extract-pkg-deps
COPY package.json /tmp/package.json
RUN jq '{name, dependencies, devDependencies}' < /tmp/package.json > /tmp/deps.json

# Install node_modules
FROM node-base AS install-modules
WORKDIR /service
COPY --from=extract-pkg-deps /tmp/deps.json package.json
COPY yarn.lock yarn.lock
RUN --mount=type=secret,id=npmrc,dst=/service/.npmrc,ro=true yarn install --pure-lockfile --non-interactive --cache-folder ./.ycache && rm -rf ./.ycache

# Build
FROM node-base AS build
WORKDIR /service
COPY --from=install-modules /service/node_modules node_modules
COPY package.json package.json
CMD yarn build
#enddogen
"
`;

exports[`generateDockerfile should generate Dockerfile for service using yarn 1`] = `
"#dogen
FROM alpine:3.12 AS jq
RUN apk add --update --no-cache jq

FROM node:latest AS node-base

# Extract minimal fields for dependencies installation
# This step avoid reinstalling node_modules due to field that is unrelated
FROM jq AS extract-pkg-deps
COPY package.json /tmp/package.json
RUN jq '{name, dependencies, devDependencies}' < /tmp/package.json > /tmp/deps.json

# Install node_modules
FROM node-base AS install-modules
WORKDIR /service
COPY --from=extract-pkg-deps /tmp/deps.json package.json
COPY yarn.lock yarn.lock
RUN yarn install --pure-lockfile --non-interactive --cache-folder ./.ycache && rm -rf ./.ycache

# Build
FROM node-base AS build
WORKDIR /service
COPY --from=install-modules /service/node_modules node_modules
COPY package.json package.json
CMD yarn build
#enddogen
"
`;
